mllib/src/main/scala/org/apache/spark/ml/feature/VectorSlicer.scala

```diff
@@ -0,0 +1,182 @@
+import org.apache.spark.ml.param.shared.{HasInputCol, HasOutputCol}
+import org.apache.spark.ml.param.{IntArrayParam, ParamMap, StringArrayParam}
+import org.apache.spark.ml.util.{Identifiable, SchemaUtils}
+import org.apache.spark.mllib.linalg._
+import org.apache.spark.sql.DataFrame
+import org.apache.spark.sql.functions._
+import org.apache.spark.sql.types.StructType
+
+/**
+ * :: Experimental ::
+ * This class takes a feature vector and outputs a new feature vector with a subarray of the
+ * original features.
+ * The subset of features can be specified with either indices ([[setSelectedIndices()]])
+ * or names ([[setSelectedNames()]]).  At least one feature must be selected.
+ * The output vector will order features with the selected indices first (in the order given),
+ * followed by the selected names (in the order given).
+ */
+@Experimental
+final class VectorSlicer(override val uid: String)
+  extends Transformer with HasInputCol with HasOutputCol {
+
+  def this() = this(Identifiable.randomUID("vectorSlicer"))
+
+  /**
+   * An array of indices to select features from a vector column.
+   * There can be no overlap with [[selectedNames]].
+   * @group param
+   */
+  val selectedIndices = new IntArrayParam(this, "selectedIndices",
+    "An array of indices to select features from a vector column." +
+      " There can be no overlap with selectedNames.", VectorSlicer.validIndices)
+
+  setDefault(selectedIndices -> Array.empty[Int])
+
+  /** @group getParam */
+  def getSelectedIndices: Array[Int] = getOrDefault(selectedIndices)
+
+  /** @group setParam */
+  def setSelectedIndices(value: Array[Int]): this.type = set(selectedIndices, value)
+
+  /**
+   * An array of feature names to select features from a vector column.
+   * These names must be specified by ML [[org.apache.spark.ml.attribute.Attribute]]s.
+   * There can be no overlap with [[selectedIndices]].
+   * @group param
+   */
+  val selectedNames = new StringArrayParam(this, "selectedNames",
+    "An array of feature names to select features from a vector column." +
+      " There can be no overlap with selectedIndices.", VectorSlicer.validNames)
+
+  setDefault(selectedNames -> Array.empty[String])
+
+  /** @group getParam */
+  def getSelectedNames: Array[String] = getOrDefault(selectedNames)
+
+  /** @group setParam */
+  def setSelectedNames(value: Array[String]): this.type = set(selectedNames, value)
+
+  /** @group setParam */
+  def setInputCol(value: String): this.type = set(inputCol, value)
+
+  /** @group setParam */
+  def setOutputCol(value: String): this.type = set(outputCol, value)
+
+  /**
+   * Slice a dense vector with an array of indices.
+   */
+  private[feature] def selectColumns(indices: Array[Int], features: DenseVector): Vector = {
+    Vectors.dense(indices.map(features.apply))
+  }
+
+  /**
+   * Slice a sparse vector with a set of indices.
+   */
+  private[feature] def selectColumns(indices: Array[Int], features: SparseVector): Vector = {
```

Comment:
Maybe we don't need two `selectColumns`. One with case matching inside should be sufficient.